samples/quickperf/src/main/java/com/google/cloud/jdbc/quickperf/QuickPerf.java

```diff
@@ -0,0 +1,497 @@
+ */
+
+package com.google.cloud.jdbc.quickperf;
+
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.security.SecureRandom;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.cli.*;
+import org.apache.commons.lang3.ArrayUtils;
+
+import com.fasterxml.jackson.core.exc.StreamReadException;
+import com.fasterxml.jackson.databind.DatabindException;
+import com.google.cloud.jdbc.quickperf.config.Config;
+import com.google.cloud.jdbc.quickperf.config.ConfigParser;
+import com.google.cloud.jdbc.quickperf.config.QueryParam;
+import com.google.cloud.spanner.Dialect;
+import com.google.cloud.spanner.jdbc.CloudSpannerJdbcConnection;
+
+import net.datafaker.Faker;
+
+public class QuickPerf extends Thread {
+    private static String BREAK_STR = "###################################################################################################";
+    private static String DEFAULT_TAG = "perftest_" + (new Random()).nextInt(300);
+
+    // TODO: make measurementfile configurable
+    private static String MEASURES_FILE_NAME = "measures.txt";
+
+    // perf measurement
+    private float measures[];
+
+    // configs
+    Config config;
+
+    private List<String> sampledValueList = new ArrayList<String>();
+
+    private int progress;
+
+    public static void main(String[] args)
+            throws SQLException, InterruptedException, StreamReadException, DatabindException, IOException {
+        Options options = new Options();
+
+        options.addOption(QuickPerf.addOption("c", "config", true, "Config File"));
+
+        CommandLineParser parser = new DefaultParser();
+        HelpFormatter formatter = new HelpFormatter();
+        CommandLine cmd = null;
+
+        ZonedDateTime testStartTimestamp = ZonedDateTime.now();
+
+        try {
+            cmd = parser.parse(options, args);
+        } catch (ParseException e) {
+            System.out.println(e.getMessage());
+            formatter.printHelp("utility-name", options);
+
+            System.exit(1);
+        }
+
+        // System.out.println("#### " + cmd.getOptionValue("config"));
+
+        Config config = ConfigParser.parseConfigFile(cmd.getOptionValue("config"));
+
+        float measures[] = new float[config.getIterations() * config.getThreads()];
+
+        List<QuickPerf> threadList = new ArrayList<QuickPerf>();
+        for (int i = 0; i < config.getThreads(); i++) {
+            QuickPerf thread = new QuickPerf(config);
+            if (config.getSamplingQuery() != null) {
+                thread.runSampling();
+            }
```

Comment:
I think that it would make sense to move this to a separate loop that is executed before the loop that starts the threads. With this setup, you potentially get threads that get a 'head start', because the later threads are spending time on the sampling query.
What I mean is that:
1. Thread 1 runs the sample query and is then started.
2. Thread 2 runs the sample query. In the meantime, thread 1 is already running the benchmark queries. This could affect the results of thread 1, depending on whether the sampling query uses more/less CPU power on Spanner.
3. Thread 3 gets the same effect as thread 2, etc.

Reply:
This will run a sampling to seed data for each thread before actually starting the thread itself. So each thread will have their own seeded data

Reply:
Yes, I understand that. But what happens here is that:
1. The application runs the sampling query for `QuickPerfRunner` number 1. That takes let's say 2 seconds.
2. `QuickPerfRunner` number 1 is started and starts executing the benchmark.
3. Step 1 is then repeated for `QuickPerfRunner` number 2. This then also takes 2 seconds.
4. Etc.... Meaning that if you have 10 threads, it takes 20 seconds before the 10th thread starts. That again means that you can have a relatively significant delay between the time that the first thread starts and the last thread starts, which again can have an impact on the results that you see (you probably also get the same effect at the end of the test run, as the first thread is likely to finish well ahead of the last thread).